test/service_manager_test.dart

```diff
@@ -0,0 +1,183 @@
+import 'dart:async';
+import 'package:devtools/eval_on_dart_library.dart';
+import 'package:devtools/globals.dart';
+import 'package:devtools/service_manager.dart';
+import 'package:devtools/vm_service_wrapper.dart';
+import 'package:flutter_tools/src/base/file_system.dart';
+import 'package:test/test.dart';
+import 'package:vm_service_lib/vm_service_lib.dart';
+
+import 'test_driver.dart';
+
+void main() {
+  group('serviceManagerTests', () {
+    FlutterRunTestDriver _flutter;
+    VmServiceWrapper service;
+
+    setUp(() async {
+      _flutter =
+          FlutterRunTestDriver(fs.directory('test/fixtures/flutter_app'));
+
+      await _flutter.run(withDebugger: true);
+      service = _flutter.vmService;
+
+      setGlobal(ServiceConnectionManager, new ServiceConnectionManager());
+
+      final completer = new Completer<Null>();
+      await serviceManager.vmServiceOpened(service, completer.future);
+    });
+
+    tearDown(() async {
+      await service.allFuturesCompleted.future;
+      _flutter.stop();
+    });
+
+    test('vmServiceOpened', () async {
+      expect(serviceManager.service, equals(service));
+      expect(serviceManager.isolateManager, isNotNull);
+      expect(serviceManager.serviceExtensionManager, isNotNull);
+      expect(serviceManager.isolateManager.isolates, isNotEmpty);
+
+      // Test should time out and fail if the following await statements do not
+      // finish with a value.
+      if (serviceManager.isolateManager.selectedIsolate == null) {
+        await serviceManager.isolateManager.onSelectedIsolateChanged
+            .firstWhere((ref) => ref != null);
+      }
+      if (serviceManager.isolateManager.flutterIsolate == null) {
+        await serviceManager.isolateManager.onFlutterIsolateChanged
+            .firstWhere((ref) => ref != null);
+      }
+    });
+
+    test('toggle boolean service extension', () async {
+      final EvalOnDartLibrary library = new EvalOnDartLibrary(
+        'package:flutter/src/rendering/debug.dart',
+        service,
+      );
+
+      StreamSubscription<ServiceExtensionState> stream;
+
+      // Initial value on the test device should be false.
+      final before = await library.eval('debugPaintSizeEnabled', isAlive: null);
+      if (before is InstanceRef) {
+        expect(before.valueAsString, 'false');
+      }
+      // Initial state in ServiceExtensionManager should be disabled with value
+      // null.
+      stream = serviceManager.serviceExtensionManager
+          .getServiceExtensionState('ext.flutter.debugPaint', null);
+      stream.onData((ServiceExtensionState state) {
+        expect(state.enabled, false);
```

Comment:
verify that onData actually got called. Otherwise you risk a test that passes for the wrong reason.
E.g. outside of this function have a counter that you increment when onData is called.
then after the function expect that the counter has value 1.

Reply:
I changed this to use a future instead of a stream subscription. This eliminated the problem with the onData callback.